dwave/optimization/src/nodes/indexing.cpp

```diff
@@ -1596,6 +1597,156 @@ std::span<const ssize_t> BasicIndexingNode::shape(const State& state) const {
                                     ndim_);
 }
 
+// ConcatenateNode ************************************************************
+
+std::vector<ssize_t> make_concatenate_shape(std::vector<ArrayNode*> array_ptrs, ssize_t axis) {
+    // One or more arrays must be given
+    if (array_ptrs.size() < 1) {
+        throw std::invalid_argument("need at least one array to concatenate");
+    }
+
+    for (auto it = std::next(array_ptrs.begin()); it != array_ptrs.end(); it++) {
```

Comment:
```c++
    for (auto it = std::next(array_ptrs.begin()), stop = array_ptrs.end(); it != stop; it++) {
```
see https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

also applies to a bunch of other places in the code.

Reply:
Im surprised issue 1 isn't optimized away by the compiler. 
Issue 2 makes perfect sense, will fix.